### PR TITLE
[ROCm] Fixes the kernel asserts API declaration mismatch error

### PR DESCRIPTION
**Pull request by pruthvistony**
This problem updates the the PR [#73040](https://github.com/pytorch/pytorch/pull/73040)

The compilation error in pyTorch with ROCm is successful with these changes when `NDEBUG` is enabled.

Solution:
For HIP we keep `__device__ __assert_fail()`
and for host side compilation we want to use the `__assert_fail()` from the glibc library.

Tested the code by compiling with below steps
```
python3 tools/amd_build/build_amd.py
python3 setup.py develop --cmake-only
cmake -DHIP_HIPCC_FLAGS_RELEASE="-DNDEBUG" build
cmake --build build
```

The UT test_fixed_cuda_assert_async is still skipped due performance overhead.

cc @jithunnair-amd

### CODE REVIEW

File: c10/macros/Macros.h [hunk 1]
```diff
@@ -332,14 +332,15 @@ constexpr uint32_t CUDA_THREADS_PER_BLOCK_FALLBACK = 256;
...
+#if defined(__ANDROID__) || defined(__APPLE__) || \
+    (defined(USE_ROCM) && ROCM_VERSION < 40100)
 // Those platforms do not support assert()
 #define CUDA_KERNEL_ASSERT(cond)
 #elif defined(_MSC_VER)
 #if defined(NDEBUG)
 extern "C" {
 C10_IMPORT
-#if defined(__CUDA_ARCH__)
+#if defined(__CUDA_ARCH__) || defined(__HIP_ARCH__) || defined(__HIP__)
```
Comment (jeffdaily) [inline c10/macros/Macros.h:343]:
```suggestion
#if defined(__CUDA_ARCH__)
```
Revert this change.  It is in the #elif _MSC_VER section and isn't relevant.  We don't build rocm + windows.

Comment (pruthvistony) [reply c10/macros/Macros.h:343]: oh yes, will update it.

**Comment by pruthvistony** [comment]
The CUDA failure
FAIL [2.540s]: test_MaxUnpool_index_errors_case1_cuda (__main__.TestNNDeviceTypeCUDA)
FAIL [2.524s]: test_MaxUnpool_index_errors_case2_cuda (__main__.TestNNDeviceTypeCUDA)
doesnt seem to be related to this change.

File: c10/macros/Macros.h [hunk 2]
```diff
@@ -362,7 +363,13 @@ extern SYCL_EXTERNAL void __assert_fail(
...
 #if (defined(__CUDA_ARCH__) && !(defined(__clang__) && defined(__CUDA__)))
 __host__ __device__
-#endif // __CUDA_ARCH__
+#elif defined(__HIP_ARCH__) || defined(__HIP__)
+__device__ __attribute__((noinline)) __attribute__((weak)) void __assert_fail(
+    const char* assertion,
+    const char* file,
+    unsigned int line,
+    const char* function);
+#endif
```
Comment (shintaro-iwasaki) [inline c10/macros/Macros.h:372]: Thanks for your PR. It's really hard to understand how this will be expanded. It seems that
```cxx
# CUDA: one declaration
__host__ __device__ void __assert_fail(...) throw() __attribute__((__noreturn__));

# HIP/ROCm: two declarations
__device__ __attribute__((noinline)) __attribute__((weak)) void __assert_fail(...);
void __assert_fail(...) throw() __attribute__((__noreturn__));
```

Suggestions:
1. Would you mind adding a comment to the code?
2. Can you clean up this `#ifdef` logic? Perhaps:
```cxx
#if CUDA
__host__ __device__
#endif
void __assert_fail(...) throw() __attribute__((__noreturn__));

#if HIP
// A comment on a reason why it is needed
__device__ __attribute__((noinline)) __attribute__((weak)) void __assert_fail(...);
#endif
```

Comment (pruthvistony) [reply c10/macros/Macros.h]: Updated the code.

**Comment by pruthvistony** [comment]
Seeing few performance regression when kernel asserts are enabled.

**Review by shintaro-iwasaki — APPROVED** [review]
Thanks. It looks good to me. Performance drop is expected if `assert()` is enabled. If this performance regression is okay for AMD (anyway the CUDA side enables this `assert()`), it's okay.

**Comment by pruthvistony** [comment]
> Thanks. It looks good to me. Performance drop is expected if `assert()` is enabled. If this performance regression is okay for AMD (anyway the CUDA side enables this `assert()`), it's okay.

Currently we are doing profiling and analysis on how the asserts are affecting few kernel by checking the generated codes. 

Can I know reason why `assert()` on device side are always enabled even in the release build?

**Comment by shintaro-iwasaki** [comment]
I don't know why `assert()` is always enabled even in the release build, but my guess is to enhance debuggability.

Is it okay for AMD to merge this (cc @jeffdaily)? If ready, please rebase this PR and ping me if tests get green.

**Comment by pruthvistony** [comment]
> I don't know why `assert()` is always enabled even in the release build, but my guess is to enhance debuggability.
> 
> Is it okay for AMD to merge this (cc @jeffdaily)? If ready, please rebase this PR and ping me if tests get green.

I have discussed with @jeffdaily , we are thinking we have to disable the asserts for ROCm, may be by adding an extra compile time guard. 

And checking on the `assert()` implementation in ROCm on HIP side we see it is different compared to CUDA. In ROCm when `assert()` is `TRUE` it communicates to host and aborts the application, but on CUDA we see it is setting the GPU to invalid state only and responsibility is on the application to abort itself. This difference in behavior is causing a significant perf overhead.

**Comment by pruthvistony** [comment]
@shintaro-iwasaki 
Please review the latest changes, I have added a compile time flag to disable kernel assert from ROCm only.

**Comment by pruthvistony** [comment]
Many Windows related jobs are failing and doesnt seem to be related to this PR.

**Comment by shintaro-iwasaki** [comment]
They seem not related to this PR, but we'd appreciate it if you could rebase this PR onto a stable one (perhaps later).

So finally what does this PR do? This PR "enables" kernel asserts but disables it by default? If so, what is a use case?

**Comment by pruthvistony** [comment]
> They seem not related to this PR, but we'd appreciate it if you could rebase this PR onto a stable one (perhaps later).
> 
> So finally what does this PR do? This PR "enables" kernel asserts but disables it by default? If so, what is a use case?

@shintaro-iwasaki , 
- I have rebase is done.
- This PR solves the __assert_fail() declaration mismatch error for ROCm. But still keep the kernel assert disabed for ROCm due to performance overhead and UT `test_cuda.test_fixed_cuda_assert_async` is skipped.

**Comment by shintaro-iwasaki** [comment]
Thanks for your explanation. Do you still want to merge this PR now if there's no particular use case, and this feature is disabled by default and will not be tested? @pruthvistony

**Comment by pruthvistony** [comment]
> Thanks for your explanation. Do you still want to merge this PR now if there's no particular use case, and this feature is disabled by default and will not be tested? @pruthvistony

@shintaro-iwasaki , 
We could like to merge this PR, since it provides a way for user to enable the asserts on ROCm, if required.

Also I am planning to update the page - https://pytorch.org/docs/stable/notes/hip.html
with detail on how they can do it.

**Comment by shintaro-iwasaki** [comment]
Thanks! Then let's merge it. This PR looks good, but we might need to rebase a few times to make all the tests green. I'd appreciate it if you can rebase this PR if we have false-positive failures.

**Comment by pruthvistony** [comment]
@shintaro-iwasaki ,
The 2 failing test from periodic/ios jobs which are not related to this change and I find them failing in other PR also. So I believe this PR has passed all the CI builds/tests and good to be merged.

**Comment by pruthvistony** [comment]
Documentation change is raised in PR - https://github.com/pytorch/pytorch.github.io/pull/1097

**Comment by pruthvistony** [comment]
> 

I had raised the PR at a wrong repo, correcting it 
https://github.com/pytorch/pytorch/pull/83220

**Comment by pruthvistony** [comment]
> Documentation change is raised in PR - [pytorch/pytorch.github.io#1097](https://github.com/pytorch/pytorch.github.io/pull/1097)

I had raised the PR at a wrong repo, correcting it
https://github.com/pytorch/pytorch/pull/83220

**Review by jeffdaily — APPROVED** [review]

**Comment by pruthvistony** [comment]
@shintaro-iwasaki ,
Can we merge this PR

**Comment by shintaro-iwasaki** [comment]
Thanks. An approval stamp from a special reviewer is needed to merge this PR (if my understanding is correct).
@ezyang or @malfet: would you please approve this PR?

**Review by malfet — APPROVED** [review]

File: c10/macros/Macros.h [hunk 3]
```diff
@@ -332,7 +332,9 @@ constexpr uint32_t CUDA_THREADS_PER_BLOCK_FALLBACK = 256;
 // CUDA_KERNEL_ASSERT checks the assertion
 // even when NDEBUG is defined. This is useful for important assertions in CUDA
 // code that would otherwise be suppressed when building Release.
-#if defined(__ANDROID__) || defined(__APPLE__) || defined(USE_ROCM)
+#if defined(__ANDROID__) || defined(__APPLE__) ||  \
+    (defined(USE_ROCM) && ROCM_VERSION < 40100) || \
+    (defined(USE_ROCM) && defined(ROCM_DISABLE_GPU_ASSERTS))
```
Comment (malfet) [inline c10/macros/Macros.h:337]: Hmm, can't we just replace this one with `|| defined(ROCM_DISABLE_GPU_ASSERTS)` and define it in CMake if version is less than 40100?

File: c10/macros/Macros.h [hunk 4]
```diff
@@ -361,22 +363,32 @@ extern SYCL_EXTERNAL void __assert_fail(
     const char* func);
 #else // __SYCL_DEVICE_ONLY__
 #if (defined(__CUDA_ARCH__) && !(defined(__clang__) && defined(__CUDA__)))
+// CUDA supports __assert_fail function which are common for both device
+// and host side code.
 __host__ __device__
-#endif // __CUDA_ARCH__
+#endif
```
Comment (malfet) [inline c10/macros/Macros.h:369]: Why this change is necessary?

File: c10/macros/Macros.h [hunk 5]
```diff
@@ -361,22 +363,32 @@ extern SYCL_EXTERNAL void __assert_fail(
...
         const char* assertion,
         const char* file,
         unsigned int line,
-        const char* function) throw()
-// We match the declaration of __assert_fail exactly how it is in glibc in case
-// parts of the program are compiled with different NDEBUG settings. Otherwise
-// we might get 'ambiguous declaration' error.
-#ifdef __GNUC__
-        __attribute__((__noreturn__))
-#endif
```
Comment (malfet) [inline c10/macros/Macros.h:377]: Can you provide a short explanation, why are you dropping this conditional? (I guess the answer is: CUDA no longer supported on Mac)

File: c10/macros/Macros.h (hunk 3 above)
Comment (pruthvistony) [reply c10/macros/Macros.h:337]: This is a new build argument to disable asserts in ROCm > 4.1. This is done due to performance regression when asserts are enabled for ROCm.

**Comment by malfet** [comment]
I don't think any special process is needed, but thank you for pining me on the issue. Buildsystem changes look fine, but can we perhaps simplify the condition? Also, as we want to disable asserts by default, I would've changed it to simple `!defined(TORCH_ENABLE_GPU_ASSERTS)`, as I assume disabling asserts would make it faster on CUDA as well(though we don't want to compile with that flag by default for CUDA as overhead is small, but do want for ROCm, as overhead is higher)

File: c10/macros/Macros.h (hunk 4 above)
Comment (pruthvistony) [reply c10/macros/Macros.h:369]: It was more of a cleanup change.

File: c10/macros/Macros.h (hunk 5 above)
Comment (pruthvistony) [reply c10/macros/Macros.h:377]: outline of final code is like
```
#if CUDA
__host__ __device__
#endif
void __assert_fail(...) throw() __attribute__((__noreturn__));

#if HIP
// A comment on a reason why it is needed
__device__ __attribute__((noinline)) __attribute__((weak)) void __assert_fail(...);
#endif
```

So essentially this `__attribute__((__noreturn__))` is kept active for CUDA case and CPU compilation code.
And build is successful is on Mac.

**Comment by pruthvistony** [comment]
> I don't think any special process is needed, but thank you for pining me on the issue. Buildsystem changes look fine, but can we perhaps simplify the condition? Also, as we want to disable asserts by default, I would've changed it to simple `!defined(TORCH_ENABLE_GPU_ASSERTS)`, as I assume disabling asserts would make it faster on CUDA as well(though we don't want to compile with that flag by default for CUDA as overhead is small, but do want for ROCm, as overhead is higher)

Ok. I will add the code to use `TORCH_ENABLE_GPU_ASSERTS` in a follow-up PR.
Thank you.

**Comment by pruthvistony** [comment]
@malfet ,
The job - "Meta Internal-Only Changes Check" failed.
I cant check the details, can you please help me on this.
